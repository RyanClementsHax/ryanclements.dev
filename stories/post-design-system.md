---
title: 'Post Design System for ryanclements.dev'
bannerAlt: 'tomatoes'
description: 'This is a post that contains all of the custom components and styles to make a post look awesome.'
---

<!-- markdownlint-disable MD029 -->

## h2 Heading

### h3 Heading

## Horizontal Rules

---

---

---

## Emphasis

**This is bold text** that begins a paragraph

_This is italic text_ that begins a paragraph

~~Strikethrough~~

## Blockquotes

> Blockquotes can also be nested...
>
> > ...by using additional greater-than signs right next to each other...
> >
> > > ...or with spaces between arrows.

## Callouts

<aside>
  I realize that using GitHub as my CMS is a little odd, but wonderful people
  like you contribute improvements to my open source content all the time and I
  appreciate that. So by keeping things in GitHub, that can continue. (Note the
  edit link at the bottom of every post).
</aside>

<aside class="info">
  This is like the one above but with the type explicitly "info"
</aside>

<aside class="info">

A `mix` of _markdown_ and <em>HTML</em>.

```js
const test = 'this is a code block in the aside'
```

`[1, 2, 3]{:js}`

</aside>

<aside class="success">
  I realize that using GitHub as my CMS is a little odd, but wonderful people
  like you contribute improvements to my open source content all the time and I
  appreciate that. So by keeping things in GitHub, that can continue. (Note the
  edit link at the bottom of every post).
</aside>

<aside class="warning">
  I realize that using GitHub as my CMS is a little odd, but wonderful people
  like you contribute improvements to my open source content all the time and I
  appreciate that. So by keeping things in GitHub, that can continue. (Note the
  edit link at the bottom of every post).
</aside>

<aside class="danger">
  I realize that using GitHub as my CMS is a little odd, but wonderful people
  like you contribute improvements to my open source content all the time and I
  appreciate that. So by keeping things in GitHub, that can continue. (Note the
  edit link at the bottom of every post).
</aside>

## Tasklist

- [ ] To do
- [x] Done

## Lists

Unordered

- Create a list by starting a line with `+`, `-`, or `*`
- Sub-lists are made by indenting 2 spaces:
  - Marker character change forces new list start:
    - Ac tristique libero volutpat at
    - Facilisis in pretium nisl aliquet
    - Nulla volutpat aliquam velit
- Very easy!

Ordered

1. Lorem ipsum dolor sit amet
2. Consectetur adipiscing elit
3. Integer molestie lorem at massa

4. You can use sequential numbers...
5. ...or keep all the numbers as `1.`

Start numbering with offset:

57. foo
1. bar

## Code

Inline `code`

Inline highlighted code `const code = "test"{:js}`

Block code "fences"

```text
Sample text here...
```

Syntax highlighting

```js
var foo = function (bar) {
  return bar++
}

console.log(foo(5))
```

Line highlighting

```rust {9, 16-18}
impl<T: 'static> AsyncRefCell<T> {
  /// Create a new `AsyncRefCell` that encapsulates the specified value.
  /// Note that in order to borrow the inner value, the `AsyncRefCell`
  /// needs to be wrapped in an `Rc` or an `RcRef`. These can be created
  /// either manually, or by using the convenience method
  /// `AsyncRefCell::new_rc()`.
  pub fn new(value: T) -> Self {
    Self {
      value: UnsafeCell::new(value),
      borrow_count: Default::default(),
      waiters: Default::default(),
      turn: Default::default(),
    }
  }

  pub fn new_rc(value: T) -> Rc<Self> {
    Rc::new(Self::new(value))
  }

  pub fn as_ptr(&self) -> *mut T {
    self.value.get()
  }

  pub fn into_inner(self) -> T {
    assert!(self.borrow_count.get().is_empty());
    self.value.into_inner()
  }
}
```

With title

```js title=next.config.js
// I'd love to convert this to .mjs but storybook-addon-next doesn't support importing mjs yet
/* eslint-disable @typescript-eslint/no-var-requires */

const withBundleAnalyzer = require('@next/bundle-analyzer')

/**
 * @type {import('next').NextConfig}
 **/
const config = {}

module.exports = () =>
  [
    withBundleAnalyzer({
      enabled: process.env.ANALYZE === 'true'
    })
  ].reduce((acc, next) => next(acc), config)
```

## Tables

| Option | Description                                                               |
| ------ | ------------------------------------------------------------------------- |
| data   | path to data files to supply the data that will be passed into templates. |
| engine | engine to be used for processing templates. Handlebars is the default.    |
| ext    | extension to be used for dest files.                                      |

Right aligned columns

| Option |                                                               Description |
| -----: | ------------------------------------------------------------------------: |
|   data | path to data files to supply the data that will be passed into templates. |
| engine |    engine to be used for processing templates. Handlebars is the default. |
|    ext |                                      extension to be used for dest files. |

## Links

[link text](http://dev.nodeca.com)

[link with title](http://nodeca.github.io/pica/demo/ 'title text!')

Autoconverted link <https://github.com/nodeca/pica>

## Images

### Figure

![Minion](minion.png)

### Image between text

This is a test for an image in between text ![Stormtroopocat](stormtroopocat.jpg 'The Stormtroopocat') cuz I need a test case

### With title

![Stormtroopocat](stormtroopocat.jpg 'The Stormtroopocat')

### With image path as footnote

Like links, Images also have a footnote style syntax

![Alt text][id]

With a reference later in the document defining the URL location:

[id]: dojocat.jpg 'The Dojocat'
