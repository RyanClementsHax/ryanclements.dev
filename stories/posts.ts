import { RenderablePostSummary } from 'lib/pages/posts'
import tomatoesPicture from 'public/tomatoes.jpg'

export const tomatoesImageData = {
  ...tomatoesPicture,
  alt: 'My family and I at a harvest festival'
}

export const postSummaries: RenderablePostSummary[] = [
  {
    slug: 'post-design-system',
    title: 'Post Design System for ryanclements.dev',
    description:
      'This is a post that contains all of the custom components and styles to make a post look awesome.',
    thumbnailSrc: tomatoesImageData
  },
  {
    slug: 'the-perfect-header-animation',
    title: 'The Perfect Header Animation',
    description:
      'Ever wondered how to create a header that animates in and out of view in perfect sync with scrolling?',
    publishedOn: 'Feb 3rd, 2022',
    thumbnailSrc: tomatoesImageData
  },
  {
    slug: 'using-webpack-loaders-for-nextjs-static-props-in-storybook',
    title: 'Using Webpack Loaders for Next.js Static Props in Storybook',
    description:
      'Using esbuild and a simple webpack loader, I found a way to just "import" the static props used by Next.js page components. You can even use node apis too!',
    publishedOn: 'Jul 14th, 2022',
    thumbnailSrc: tomatoesImageData
  }
]
