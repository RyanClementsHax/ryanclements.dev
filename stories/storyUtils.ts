import { Story, Parameters } from '@storybook/react'
import { Theme } from 'components/theme'
import merge from 'just-merge'

export interface StoryModifier {
  <T>(story: Story<T>): Story<T>
}

export const withNoop: StoryModifier = story => story

export const compose =
  (...params: StoryModifier[]): StoryModifier =>
  story =>
    params.reverse().reduceRight((y, fn) => fn(y), story)

export const asCopy: StoryModifier = story => story.bind({})

export const withParams: (params: Parameters) => StoryModifier =
  params => story => {
    story.parameters = merge(story.parameters || {}, params)
    return story
  }

export const withDarkTheme: StoryModifier = withParams({ theme: Theme.dark })

export const withMobile: StoryModifier = withParams({
  viewport: {
    defaultViewport: 'mobile1'
  }
})

export const withFigmaUrl: (url: string) => StoryModifier = url =>
  withParams({
    design: {
      type: 'figma',
      url
    }
  })

export interface DefaultStoryConfig {
  figmaUrl?: string
}

export const withDefaults: (config?: DefaultStoryConfig) => StoryModifier =
  config =>
    compose(asCopy, config?.figmaUrl ? withFigmaUrl(config.figmaUrl) : withNoop)

export const createDefaultStories = <T>(
  template: Story<T>,
  {
    base,
    mobile,
    darkTheme,
    darkThemedMobile
  }: {
    base?: DefaultStoryConfig
    mobile?: DefaultStoryConfig
    darkTheme?: DefaultStoryConfig
    darkThemedMobile?: DefaultStoryConfig
  } = {}
): {
  Base: Story<T>
  Mobile: Story<T>
  DarkTheme: Story<T>
  DarkThemedMobile: Story<T>
} => ({
  Base: withDefaults(base)(template),
  Mobile: compose(withDefaults(mobile), withMobile)(template),
  DarkTheme: compose(withDefaults(darkTheme), withDarkTheme)(template),
  DarkThemedMobile: compose(
    withDefaults(darkThemedMobile),
    withMobile,
    withDarkTheme
  )(template)
})
