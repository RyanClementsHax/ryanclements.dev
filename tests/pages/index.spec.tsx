import React from 'react'
import { render, act } from 'tests/testUtils'
import { axe } from 'jest-axe'
import { Primary } from 'stories/pages/index.stories'

describe('Index page', () => {
  let container: Element, asFragment: () => DocumentFragment

  beforeEach(() => {
    ;({ container, asFragment } = render(<Primary />))
  })

  it('has no axe violations', async () => {
    await act(async () => expect(await axe(container)).toHaveNoViolations())
  })
  it('matches snapshot', () => {
    expect(asFragment()).toMatchSnapshot()
  })
})
