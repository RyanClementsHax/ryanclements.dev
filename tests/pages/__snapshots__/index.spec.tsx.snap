// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index page matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col items-center justify-center min-h-screen py-2"
  >
    <main
      class="flex flex-col items-center justify-center w-full flex-1 px-20 text-center"
    >
      <h1
        class="text-6xl font-bold"
      >
        Welcome to my site!
      </h1>
      <p
        class="mt-3 text-2xl"
      >
        It is currently under construction 🚧👷‍♂️👷‍♀️
      </p>
    </main>
    <footer
      class="flex items-center justify-center w-full h-24 border-t"
    >
      <a
        class="flex items-center justify-center"
        href="https://vercel.com?utm_source=create-next-app&utm_medium=default-template&utm_campaign=create-next-app"
        rel="noopener noreferrer"
        target="_blank"
      >
        Powered by
        <span
          class="h-4 ml-2 mr-2"
        >
          <div
            style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
          >
            <div
              style="box-sizing: border-box; display: block; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNzAiIGhlaWdodD0iMTYiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
                style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
              />
            </div>
            <img
              alt="Vercel Logo"
              data-nimg="intrinsic"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </div>
        </span>
        and ✝️
      </a>
    </footer>
  </div>
</DocumentFragment>
`;
